### PR TITLE
Implement makeSplineApprox for edges and faces

### LINKED ISSUE #682: Expose GeomAPI_PointsToBSplineSurface
And maybe implement `parametricSurface`. While we are at it, let's expose `GeomAPI_PointsToBSpline` and use that in `parametricCurve`

Related to #675

### PR DESCRIPTION
**Pull request by adam-urbanczyk**
Will resolve #682 

- [x] implement `Edge.makeSplineApprox`
- [x] implement `Face.makeSplineApprox`
- [x] rework `parametricCurve`
- [x] implement `parametricSurface`
- [x] implement `Edge.close`
- [x] add splineAppox to the CQ API
- [x] `BRepAlgoAPI_Splitter` and rework Workplane.split

Also related to #675

### CODE REVIEW

**Comment by marcus7070** [comment]
Will `Shape.split` let me split a Wire with a Vertex?

I have needed to do that before, when making a ruled surface between a Wire with 4 edges and a spline with one edge. I needed to add 3 more vertices at certain locations on the spline. It was awkward and being able to split wires would be great improvement!

**Comment by adam-urbanczyk** [comment]
> Will `Shape.split` let me split a Wire with a Vertex?
> 
> I have needed to do that before, when making a ruled surface between a Wire with 4 edges and a spline with one edge. I needed to add 3 more vertices at certain locations on the spline. It was awkward and being able to split wires would be great improvement!

I did a quick check on it seems to work.

**Comment by marcus7070** [comment]
I pushed two more tests. Feel free to revert if you don't want them, it was just easier to write the code than suggest them in comments.

File: cadquery/occ_impl/shapes.py [hunk 1]
```diff
@@ -1379,6 +1397,20 @@ def _geomAdaptorH(self) -> Tuple[BRepAdaptor_Curve, BRepAdaptor_HCurve]:
...
+        rv: Union[Wire, Edge]
+
+        if not self.IsClosed():
+            e = Edge.makeLine(self.endPoint(), self.startPoint())
+            rv = Wire.assembleEdges((self, e))
+        else:
+            rv = self
+
+        return rv
+
```
Comment (marcus7070) [inline cadquery/occ_impl/shapes.py:1413]: Would this go better on Mixin1D so it could close Wires as well?

**Review by marcus7070 — COMMENTED** [review]
I've still got more to look at here, I'll try to get back to it later today.

**Comment by marcus7070** [comment]
@adam-urbanczyk pushed some 500 lines of code without needing the "black fix" commit, I push like 15 LOC and mess up the streak. :facepalm:

**Comment by marcus7070** [comment]
I don't think we test splitting with multiple tool Shapes?

Thinking about that some more, I'm ok with not testing it. I can't actually think of when I would ever split with multiple tools, I think it's a bit of an edge case.

**Review by marcus7070 — APPROVED** [review]

File: cadquery/cq.py [hunk 2]
```diff
@@ -284,28 +292,64 @@ def split(self: T, keepTop: bool = False, keepBottom: bool = False) -> T:
...
+                        "keepBottom", Parameter.POSITIONAL_OR_KEYWORD, default=False
+                    ),
+                )
+            )
+
+            bound_args = sig.bind(*args, **kwargs)
+            bound_args.apply_defaults()
+
+            keepTop = bound_args.arguments["keepTop"]
+            keepBottom = bound_args.arguments["keepBottom"]
```
[the remark this answers is not in the document]
Comment (adam-urbanczyk) [reply cadquery/cq.py:327]: Because `keepTop` and `keepBottom` can be passed as non-keyword arguments in *args.

File: cadquery/cq.py [hunk 3]
```diff
@@ -1844,31 +1895,146 @@ def spline(
...
+        for i in range(N + 1):
+            allPoints.append(
+                self._toVectors(
+                    (
+                        func(start + diff * i / N, start + diff * j / N)
+                        for j in range(N + 1)
+                    ),
+                    False,
+                )
+            )
```
[the remark this answers is not in the document]
Comment (adam-urbanczyk) [reply cadquery/cq.py]: Looks good

File: cadquery/occ_impl/shapes.py (hunk 1 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:1413]: I don't think that this part is generic - I think it is nice to keep closed edges as `Edge`. I could add a Wire specific close to the Wire class.

Comment (marcus7070) [reply cadquery/occ_impl/shapes.py:1413]: > I don't think that this part is generic - I think it is nice to keep closed edges as Edge.

If self is an Edge, and `self.IsClosed()` is True, it will just return self, right? Edges will stay Edges?

> I could add a Wire specific close to the Wire class.

I have no need for it, just thought I saw an opportunity to generalise this with no extra code.

Comment (adam-urbanczyk) [reply cadquery/occ_impl/shapes.py:1413]: Again, what's the point of using a Mixing if you are dispatching in it on the specific type of self.

Comment (marcus7070) [reply cadquery/occ_impl/shapes.py:1413]: Ah, sorry, I tested moving this to `Mixin1D.close` locally, but I must have stuffed up my mypy check, it was all green. I see the problem now.

File: cadquery/cq.py [hunk 4]
```diff
@@ -284,28 +292,64 @@ def split(self: T, keepTop: bool = False, keepBottom: bool = False) -> T:
...
+            top = solid.cut(bottomCutBox)
+            bottom = solid.cut(topCutBox)
+
+            if keepTop and keepBottom:
+                # Put both on the stack, leave original unchanged.
+                rv = [top, bottom]
             else:
-                return self.newObject([bottom])
+                # Put the one we are keeping on the stack, and also update the
+                # context solidto the one we kept.
```
Comment (jmwright) [inline cadquery/cq.py:346]: `solidto` should be `solid to` here. I'm guessing it's a left-over from comments when this was first implemented.

File: cadquery/occ_impl/shapes.py [hunk 5]
```diff
@@ -1537,6 +1569,45 @@ def makeSpline(
...
+        smoothing: Optional[Tuple[float, float, float]] = None,
+        minDeg: int = 1,
+        maxDeg: int = 6,
+    ) -> "Edge":
+        """
+        Approximate a spline through the provided points.
+
+        :param listOfVector: a list of Vectors that represent the points
+        :param tol: tolerance of the algorithm (consult OCC documentation).
+        :param smoothing: optional tuple of 3 weigths use for variational smoothing (default: None)
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:1586]: `weigths` here should be `weights`

File: cadquery/occ_impl/shapes.py [hunk 6]
```diff
@@ -2022,6 +2093,47 @@ def makeFromWires(
...
+        smoothing: Optional[Tuple[float, float, float]] = None,
+        minDeg: int = 1,
+        maxDeg: int = 3,
+    ) -> "Face":
+        """
+        Approximate a spline surface through the provided points.
+
+        :param points: a 2D list of Vectors that represent the points
+        :param tol: tolerance of the algorithm (consult OCC documentation). 
+        :param smoothing: optional tuple of 3 weigths use for variational smoothing (default: None)
```
Comment (jmwright) [inline cadquery/occ_impl/shapes.py:2110]: `weights` has transposed characters here too

File: tests/test_cadquery.py [hunk 7]
```diff
@@ -4387,3 +4394,119 @@ def test2Dchamfer(self):
...
+        w = Edge.makeSplineApprox(pts).close()
+
+        self.assertTrue(e3.IsClosed())
+        self.assertTrue(w.IsClosed())
+
+        # Workplane method
+
+        w1 = Workplane().splineApprox(pts)
+        w2 = Workplane().splineApprox(pts, forConstruction=True)
+        w3 = Workplane().splineApprox(pts, makeWire=True,)
```
Comment (jmwright) [inline tests/test_cadquery.py:4429]: Trailing comma here that doesn't hurt anything.

**Review by jmwright — APPROVED** [review]
Thanks @adam-urbanczyk this was a lot of work. I just had a few minor comments.

File: cadquery/cq.py (hunk 4 above)
Comment (adam-urbanczyk) [reply cadquery/cq.py]: Solved

**Comment by adam-urbanczyk** [comment]
OK, I think this is ready to be merged.

**Comment by jmwright** [comment]
@adam-urbanczyk Thanks again!
